aten/src/ATen/core/jit_type.h

```diff
@@ -324,25 +324,50 @@ inline c10::optional<T> merge_primitive(
   return c10::optional<T>{};
 }
 
-// `VaryingShape` tracks if individual dimensions or a rank vary across
-// profiled runs. A *varying* or *dynamic* dimension is expressed as
-// an empty c10::optional in `sizes_`. If a rank is dynamic, the entire
-// `sizes_` becomes the empty optional.
+struct CAFFE2_API ShapeSymbol {
+    ShapeSymbol(int64_t val, bool statik = true): value_(val), statik_(statik) {}
+
+    explicit operator int() const { return value_; }
+    ShapeSymbol(const ShapeSymbol& ss) {
+      statik_ = ss.statik_;
+      value_ = ss.value_;
+    }
+    ShapeSymbol(): value_(-1), statik_(false) {}
+    static ShapeSymbol getInvalidSymbol() {return ShapeSymbol(); }
+    int64_t value_;
+    bool statik_;  
+    bool operator ==(const ShapeSymbol &b) const { return value_ == b.value_ && statik_ == b.statik_; }
+    // dynamic shapes are typically negative
+    bool operator <(const ShapeSymbol &b) const { return static_cast<int>(statik_) < static_cast<int>(b.statik_) || value_ < b.value_; }
+};
```

Comment:
I would expect to have an API for this that was something like:

```
auto s = Shape::fromConstantSize(3);
auto n = ShapeScope::fresh(); // new shape in the scope (i.e. the function being profiled)
std::optional<int> concrete_value = s.static_size();
```

Reply:
I wanted to defer an API clean-up until we have the entire E2E pipeline working for dynamic shapes, but it's indeed pretty confusing in its current state, so I cleaned it up.

> auto s = Shape::fromConstantSize(3);

added `fromStaticSize`

> auto n = ShapeScope::fresh(); // new shape in the scope (i.e. the function being profiled)

Unfortunately, this needs to be a bit more complex than this since `ShapeScope` needs to persist across multiple profiles of the same function. For the first iteration, I just added `newSymbol` which generates a fresh symbol unique globally across all the functions.

> std::optional<int> concrete_value = s.static_size();

Added `static_size` returning `int` and `is_static`